Migrating Quicktest to PyTest and adding dockerized self hosted runner

taha-abdullah opened the pull request:
1. added tests for:
   - segmentation stats
   - image data

2. migrated tests from unittest to pytest framework

3. changed fastsurfer to run multiple subjects

4. added new workflow (quicktest_runner) to run on dockerized self-hosted runner

taha-abdullah committed:
PR Review fixes:

- added support for multiple test subjects and parameterization in all tests

- reorganized file structure

- separate stats files for each aseg and aparc+DKT

- moving fixtures and common functions to common.py and conftest.py

quicktest_runner.yaml:

- removed check-output job
- added THIS_RUN_DIR to seperate different runs

taha-abdullah commented:
@dkuegler requesting review

dkuegler commented on .github/workflows/quicktest.yaml:
```diff
@@ -1,36 +1,41 @@
     - name: Set up Singularity
       uses: eWaterCycle/setup-singularity@v7
       with:
         singularity-version: 3.8.3
```
Can you check, if this can be easily changed to apptainer (the command line interface is extremely similar)
Alternatively, singularity should be updated .... it is currently 4.1+.

dkuegler commented on .github/workflows/quicktest.yaml:
```diff
@@ -1,36 +1,41 @@
+    runs-on: self-hosted
     needs: prepare-job
     steps:
     - name: Build Docker Image and convert to Singularity
```
There should be checks for whether RUNNER_SINGULARITY_IMGS and PYTHON_PATH are defined.
cd into PYTHON_PATH makes no sense, PYTHONPATH can be empty or multiple paths delimited by ":". I think you mean FASTSURFER_HOME or similar here?

dkuegler commented on .github/workflows/quicktest.yaml:
```diff
@@ -1,36 +1,41 @@
-name: FastSurfer Singularity
+name: quicktest
```
add author, created on
functionality, usage, etc.

dkuegler commented on .github/workflows/quicktest_runner.yaml:
```diff
@@ -0,0 +1,103 @@
+name: quicktest-docker
```
add author, created on
functionality, usage, etc.
A list of "expected/used" environment variables. Also things like FreeSurfer license (which should be stored in the environment variable FS_LICENSE, see https://github.com/Deep-MI/FastSurfer/blob/ece9f3578e40231887cb08442b2862492f606228/run_fastsurfer.sh#L308 ).

dkuegler commented on .github/workflows/quicktest_runner.yaml:
```diff
@@ -0,0 +1,103 @@
+      uses: mamba-org/setup-micromamba@v1
+      with:
+        environment-file: env/fastsurfer.yml
+        micromamba-root-path: /home/micromamba
```
should this not also be $MAMBAPATH?

dkuegler commented on .github/workflows/quicktest_runner.yaml:
```diff
@@ -0,0 +1,103 @@
+    - name: Cache Python packages
+      uses: actions/cache@v3
+      with:
+        path: ~/.cache/pip
```
for consistentcy sake, I think this should probably be /home/.cache/pip or consistently use ~ or $HOME instead of /home. I would even say it might be better to call this $WORK (which can just be /home in practice)

dkuegler commented on .github/workflows/quicktest_runner.yaml:
```diff
@@ -0,0 +1,103 @@
+      run: |
+        if [ ! -f "$MAMBAPATH" ]; then
+          echo "FastSurfer environment does not exist. Creating FastSurfer Environment..."
+          echo "PROCEED=true" >> $GITHUB_OUTPUT
```
I think I would call this variable "HAS_MAMBA" instead of "PROCEED"

dkuegler commented on .github/workflows/quicktest_runner.yaml:
```diff
@@ -0,0 +1,103 @@
+  run-fastsurfer:
+    runs-on: self-hosted
+    needs: checkout
+    steps:
```
there should probably be a simple step here that checks the required environment variables and settings (such as FreeSurfer license, FreeSurfer installation, etc.)

dkuegler commented on .github/workflows/quicktest_runner.yaml:
```diff
@@ -0,0 +1,103 @@
+        export FREESURFER_HOME=/freesurfer
+        source $FREESURFER_HOME/SetUpFreeSurfer.sh
+        git config --global --add safe.directory \
+        /home/actions-runner/_work/FastSurfer/FastSurfer
```
consistency of /home ... should /home/actions-runner/_work/FastSurfer/FastSurfernot just be "."?

dkuegler commented on .github/workflows/quicktest_runner.yaml:
```diff
@@ -0,0 +1,103 @@
+    - name: Cache Python packages
+      uses: actions/cache@v3
+      with:
+        path: ~/.cache/pip
```
/home

dkuegler commented on .github/workflows/quicktest_runner.yaml:
```diff
@@ -0,0 +1,103 @@
+        restore-keys: |
+          ${{ runner.os }}-pip-
+    - name: Run pytest
+      run: python -m pytest test/quicktest
```
probably should get args with directories, e.g. $RUNNER_FS_MRI_DATA/subjects_list.txt, $RUNNER_FS_OUTPUT, etc. (it is better to be explicit here)

dkuegler commented on pyproject.toml:
```diff
@@ -82,9 +82,14 @@ style = [
 all = [
     'fastsurfer[doc]',
     'fastsurfer[style]',
+    'fastsurfer[quicktest]',
```
one is quicktest/one is test... is inconsistent

dkuegler commented on pyproject.toml:
```diff
@@ -82,9 +82,14 @@ style = [
     'ruff',
 ]
+test = [
+    'pandas',
```
pandas is already in the core dependencies (see line 39)

dkuegler commented on requirements.txt:
```diff
@@ -207,4 +207,4 @@ yacs==0.1.8
 
 # The following packages are considered to be unsafe in a requirements file:
-# setuptools
+# setuptools
```
missing newline at end, should probably be reverted.

dkuegler commented on test/quick_test/data/stats.yaml:
```diff
@@ -0,0 +1,23 @@
+ lhSurfaceHoles: 10
+ rhSurfaceHoles: 10
+ SurfaceHoles: 10
+ eTIV: 10
```
missing newline at end

dkuegler commented on test/quick_test/data/stats.yaml:
```diff
@@ -0,0 +1,23 @@
+default_threshold: 0.01
+
+thresholds:
```
this should probably have a comment here that indicates these thresholds are relative and percent, but really I think it makes more sense if the value are 0.1 if they are supposed to be 10%.

dkuegler commented on test/quick_test/data/errors.yaml:
probably better to call this file `logfile.keywords.yaml`

dkuegler commented on test/quick_test/data/stats.yaml:
probably better to call this file `data/thresholds/aseg.stats.yaml` and have one file per statsfile that we check, so also have `aparc+DKT.stats.yaml`.

dkuegler commented on test/quick_test/test_errors.py:
test_keywords_in_logfiles.py ?

dkuegler commented on test/quick_test/test_images.py:
```diff
@@ -0,0 +1,219 @@
+    return image
+
+
+def compute_dice_score(test_data, reference_data):
```
can this not be imported from https://github.com/Deep-MI/FastSurfer/blob/ece9f3578e40231887cb08442b2862492f606228/CerebNet/utils/metrics.py#L170

dkuegler commented on test/quick_test/test_images.py:
```diff
@@ -0,0 +1,219 @@
+    """
+
+    # Labels
+    labels = [2, 3, 4, 5, 7, 8, 10, 11, 12, 13, 14, 15,
```
this should be in a yaml file or in fixtures

dkuegler commented on test/quick_test/test_images.py:
```diff
@@ -0,0 +1,219 @@
+    headers = [test_image.header, reference_image.header]
+
+    # Check the image header
+    assert nibabel.cmdline.diff.get_headers_diff(headers) == OrderedDict(), "Image headers do not match"
```
```suggestion
    header_diff = nibabel.cmdline.diff.get_headers_diff(headers)
    assert header_diff == OrderedDict(), f"Image headers do not match: {header_diff}"
```

dkuegler commented on test/quick_test/test_images.py:
```diff
@@ -0,0 +1,219 @@
+    print("Image headers are correct")
+
+
+def test_seg_data(sample_subjects_dir, sample_subject_id, sample_reference_id):
```
parameters should probably be subjects_dir, reference_subjects_dir and subect_id.
subject_id should then probably be a fixture with params.

We also talked about some general ideas (e.g. moving data to fixtures)

taha-abdullah replied:
the reference subjects are inside of subjects_dir

dkuegler commented on test/quick_test/test_images.py:
```diff
@@ -0,0 +1,219 @@
+
+    # Check the dice score
+    for score in dice_score:
+        assert score == 1, f"Dice score is not 1 for f{np.where(dice_score == score)[0][0]}"
```
multiple asserts per call, see if this can be made into one assert (either by splitting into multiple functions or by combining the checks via `any()`.

dkuegler commented on test/quick_test/test_stats.py:
```diff
@@ -0,0 +1,153 @@
+    return thresholds
+
+
+def read_measure_stats(file_path):
```
probably use read_measurefile (from https://github.com/Deep-MI/FastSurfer/pull/474)

dkuegler commented on test/quick_test/test_stats.py:
```diff
@@ -0,0 +1,153 @@
+        variation = ((stats_data[1].get(struct) / ref_data[1].get(struct)) - 1)
+        # If threshold exists for struct, use it, else use default threshold
+        if threshold := thresholds.get(struct):
+            assert variation <= threshold, f"Variation of {struct} is greater than threshold."
```
one assert per function

dkuegler commented on test/quick_test/test_stats.py:
```diff
@@ -0,0 +1,153 @@
+            if ref_table.loc[i, j] == 0:
+                continue
+            variation = ((test_table.loc[i, j] / ref_table.loc[i, j]) - 1)
+            assert variation <= threshold, f"Variation of {j} in the table is greater than threshold."
```
one assert per function

dkuegler commented on .github/workflows/quicktest_runner.yaml:
```diff
@@ -1,6 +1,10 @@
+  pull_request:
+    branches:
+      - dev
+      - main
```
we do not have main, just dev and stable

taha-abdullah commented:
@dkuegler requesting review after suggested fixes

dkuegler commented on .github/workflows/quicktest.yaml:
```diff
@@ -1,90 +1,111 @@
+#                FastSurfer repo, sets up Python, builds a Singularity image, runs FastSurfer on sample MRI data, and
+#                runs pytest to check if the results are acceptable
+# Usage: This workflow is triggered on a push or pull-request to the dev and main branch of DeepMI/FastSurfer.
+#        It can also be triggered manually with workflow-dispatch
```
```suggestion
# Usage: This workflow is exclusively triggered manually with workflow-dispatch in DeepMI/FastSurfer.
```

dkuegler commented on .github/workflows/quicktest.yaml:
```diff
@@ -1,90 +1,111 @@
+    - name: Build Docker Image and convert to Apptainer
       run: |
-        cd $RUNNER_SINGULARITY_IMGS
+        cd $RUNNER_FASTSURFER_IMGS
```
```suggestion
        if [[ -z "$RUNNER_FASTSURFER_IMGS" ]] || [[ ! -d "$RUNNER_FASTSURFER_IMGS" ]] ; then echo "RUNNER_FASTSURFER_IMGS is empty or does not exist: $RUNNER_FASTSURFER_IMGS" ; exit 1 ; fi
        cd $RUNNER_FASTSURFER_IMGS
```

dkuegler commented on .github/workflows/quicktest.yaml:
```diff
@@ -1,90 +1,111 @@
           echo "SIF File does not exist. Building file."
-          PYTHONPATH=$PYTHONPATH
-          cd $PYTHONPATH
+          cd $FASTSURFER_HOME
           python3 Docker/build.py --device cuda --tag fastsurfer_gpu:cuda
```
```suggestion
          if [[ -z "$FASTSURFER_HOME" ]] || [[ ! -d "$FASTSURFER_HOME" ]] ; then echo "FASTSURFER_HOME is empty or does not exist: $FASTSURFER_HOME" ; exit 1; fi
          python3 $FASTSURFER_HOME/Docker/build.py --device cuda --tag fastsurfer_gpu:cuda
```

dkuegler commented on .github/workflows/quicktest.yaml:
```diff
@@ -1,90 +1,111 @@
-          cd $RUNNER_SINGULARITY_IMGS
-          singularity build --force fastsurfer-gpu.sif docker-daemon://fastsurfer_gpu:cuda
+          cd $RUNNER_FASTSURFER_IMGS
+          apptainer build --force fastsurfer-gpu.sif docker-daemon://fastsurfer_gpu:cuda
```
```suggestion
          apptainer build --force fastsurfer-gpu.sif docker-daemon://fastsurfer_gpu:cuda
```

dkuegler commented on .github/workflows/quicktest.yaml:
```diff
@@ -1,90 +1,111 @@
         else
           echo "File already exists"
-          cd $PYTHONPATH
+          cd $FASTSURFER_HOME
```
```suggestion
```

dkuegler commented on .github/workflows/quicktest.yaml:
```diff
@@ -1,90 +1,111 @@
     - name: Run FastSurfer
       run: |
-        singularity exec --nv \
+        cd $RUNNER_FS_OUTPUT
```
```suggestion
         if [[ -z "$RUNNER_FS_OUTPUT" ]] || [[ ! -d "$RUNNER_FS_OUTPUT" ]] ; then echo "RUNNER_FS_OUTPUT is empty or does not exist: $RUNNER_FS_OUTPUT" ; exit 1; fi
         if [[ -z "$RUNNER_FS_LICENSE" ]] || [[ ! -f "$RUNNER_FS_LICENSE" ]] ; then echo "RUNNER_FS_LICENSE is empty or does not exist: $RUNNER_FS_LICENSE" ; exit 1; fi
         if [[ -z "$RUNNER_FS_MRI_DATA" ]] || [[ ! -d "$RUNNER_FS_MRI_DATA" ]] ; then echo "RUNNER_FS_MRI_DATA is empty or does not exist: $RUNNER_FS_MRI_DATA" ; exit 1; fi
         if [[ ! -f "$RUNNER_FS_MRI_DATA/subject_list.txt" ]] ; then echo "workflow expects a subject_list.txt in RUNNER_FS_MRI_DATA: $RUNNER_FS_MRI_DATA" ; exit 1; fi
         cd $RUNNER_FS_OUTPUT
```

dkuegler commented on .github/workflows/quicktest.yaml:
```diff
@@ -1,90 +1,111 @@
                       --env FASTSURFER_HOME=/fastsurfer-dev \
                       -B $RUNNER_FS_MRI_DATA:/data \
                       -B $RUNNER_FS_OUTPUT:/output \
                       -B $RUNNER_FS_LICENSE:/fs_license \
```
```suggestion
                      -B $RUNNER_FS_LICENSE:/fs_license/.license \
```

dkuegler commented on .github/workflows/quicktest_runner.yaml:
```diff
@@ -0,0 +1,150 @@
+      # Check if the Environment Variables used in further steps are present
+    - name: Check Environment Variables
+      run: |
+        REQUIRED_ENV_VARS=(
+          "MAMBAPATH"
+          "MAMBAROOT"
+          "RUNNER_FS_OUTPUT"
+          "RUNNER_FS_MRI_DATA"
+          "FREESURFER_HOME"
+          "FS_LICENSE"
+        )
+        
+        for VAR_NAME in "${REQUIRED_ENV_VARS[@]}"; do
+          if [ -z "${!VAR_NAME}" ]; then
+            echo "Error: Required environment variable $VAR_NAME is not set"
+            exit 1
+          fi
+        done
```
:+1:

dkuegler commented on .github/workflows/quicktest_runner.yaml:
```diff
@@ -0,0 +1,150 @@
+        git config --global --add safe.directory \
+        .
+        ./brun_fastsurfer.sh --subject_list $RUNNER_FS_MRI_DATA/subjects_list.txt \
+                             --sd $RUNNER_FS_OUTPUT \
```
```suggestion
                             --sd $THIS_RUN_OUTDIR --license $FS_LICENSE\
```
Currently, the license if implicitly grabbed from FREESURFER_HOME via `source $FREESURFER_HOME/SetUpFreeSurfer.sh`.

dkuegler commented on test/quick_test/test_errors_in_logfiles.py:
```diff
@@ -0,0 +1,140 @@
+    errors : list
+        List of errors.
+    whitelist : list
+        List of whitelisted errors.
```
```suggestion
    errors : list[str]
        List of errors.
    whitelist : list[str]
        List of whitelisted errors.
```

dkuegler commented on test/quick_test/test_errors_in_logfiles.py:
```diff
@@ -0,0 +1,140 @@
+
+    Returns
+    -------
+    test_paths : list
```
```suggestion
    test_paths : list[Path]
```

dkuegler commented on test/quick_test/test_errors_in_logfiles.py:
```diff
@@ -0,0 +1,140 @@
+
+    Returns
+    -------
+    log_files : list
```
```suggestion
    log_files : list[Path]
```

dkuegler commented on test/quick_test/test_errors_in_logfiles.py:
```diff
@@ -0,0 +1,140 @@
+@pytest.fixture
+def load_errors():
+    """
+    Load the errors and whitelist from the given file path.
```
```suggestion
    Load the lists of error and whitelist strings from ./data/logfile.errors.yaml.
```

dkuegler commented on test/quick_test/test_errors_in_logfiles.py:
```diff
@@ -0,0 +1,140 @@
+    """
+
+    subjects_dir = os.environ["SUBJECTS_DIR"]
+    subject_id = os.environ["SUBJECT_ID"]
```
should `subjects_dir` and `subject_id` not be fixtures?

dkuegler commented on test/quick_test/test_errors_in_logfiles.py:
```diff
@@ -0,0 +1,140 @@
+    return test_paths
+
+
+def load_log_files(test_dir: str):
```
should this function not also be a fixture?

```suggestion
def load_log_files(test_dir: Path):
```

I think I would prefer, if strings that are paths are usually `Path` objects.

dkuegler commented on test/quick_test/test_errors_in_logfiles.py:
```diff
@@ -0,0 +1,140 @@
+        for line in file:
+            filename = line.strip()
+            logger.debug(f"Processing {filename}")
+            test_file = os.path.join(subjects_dir, filename)
```
```suggestion
            test_file = Path(subjects_dir) / filename
```

dkuegler commented on test/quick_test/test_errors_in_logfiles.py:
```diff
@@ -0,0 +1,140 @@
+
+
+@pytest.mark.parametrize("test_file", load_test_files())
+def test_errors(test_file: str,
```
```suggestion
def test_errors(test_file: Path,
```

dkuegler commented on test/quick_test/test_errors_in_logfiles.py:
```diff
@@ -0,0 +1,140 @@
+
+    Parameters
+    ----------
+    test_file : str
```
```suggestion
    test_file : Path
```

dkuegler commented on test/quick_test/test_errors_in_logfiles.py:
```diff
@@ -0,0 +1,140 @@
+        If any of the keywords are in the log files.
+    """
+
+    log_files = load_log_files(test_file)
```
with `load_log_files` a fixture, should this not move into the function parameters?

taha-abdullah replied:
load_log_files is not a fixture

dkuegler commented on test/quick_test/test_errors_in_logfiles.py:
```diff
@@ -0,0 +1,140 @@
+
+    # Check if any of the keywords are in the log files
+    for log_file in log_files:
+        rel_path = log_file.relative_to(os.environ["SUBJECTS_DIR"])
```
As a consequence of use of fixtures, this should probably be removed.

dkuegler commented on test/quick_test/test_file_existence.py:
```diff
@@ -1,71 +1,132 @@
-    TestFileExistence.folder_path = args.folder_path
+    # Check if each file in the reference list exists in the test list
+    for file in reference_files:
+        assert file in test_files, f"File '{file}' does not exist."
```
This is a for loop over assert statements

dkuegler commented on test/quick_test/test_file_existence.py:
```diff
@@ -1,71 +1,132 @@
-            self.assertIn(file, self.filenames, f"File '{file}' does not exist in the folder.")
 
-        print("All files present")
+def get_files_from_yaml(file_path: str):
```
This function seems to be never used.

I think there should be a fixture called `required_files` or similar, which should use this function... I would probably go a step further and just remake this function into this fixture.

dkuegler commented on test/quick_test/test_file_existence.py:
```diff
@@ -1,71 +1,132 @@
+@pytest.mark.parametrize("test_file", load_test_files())
+def test_file_existence(test_file: str,
+                        reference_file: str):
+    """
```
I kind of think this file should have 3 parameters, all of them fixtures:
1. `subject_id` (this can be a list/multiple -- get dirs in `os.environ["SUBEJCTS_DIR"]` --- also used by other tests in other files)
2. `subjects_dir` (this is just a fixture that uses `os.environ["SUBJECTS_DIR"]` --- also used by other tests in other files)
3. `file_to_check` (this can be a list/multiple -- loaded from yaml file)

The function body then just checks if the file exists for the (`subjects_dir/subject_id/file_to_check`)

dkuegler commented on test/quick_test/test_file_existence.py:
```diff
@@ -1,71 +1,132 @@
+    return files
+
+
+def get_files_from_folder(folder_path: str):
```
I am not really sure how this function "helps".

dkuegler reviewed (changes requested):
I have not really checked test_images or test_stats yet, changes from other files should probably be applied there first too .

dkuegler commented:
Please clean up/squash the git history a bit.

dkuegler commented on test/quick_test/conftest.py:
```diff
@@ -0,0 +1,123 @@
+
+
+@pytest.fixture
+def reference_id():
+    return os.environ["REFERENCE_ID"]
```
Is this needed? should those not just be read from the subjects_list?

dkuegler commented on .github/workflows/quicktest_runner.yaml:
```diff
@@ -0,0 +1,154 @@
+    steps:
+    - name: Check FastSurfer Output Directory
+      run: |
+        if [ -d "$RUNNER_FS_OUTPUT/subject1" ]; then
+          echo "FastSurfer output directory for subject1 exists. Finishing job..."
+          echo "HAS_FASTSURFER=false" >> $GITHUB_OUTPUT
+          exit 0
+        else
+          echo "FastSurfer output directory for subject1 does not exist. Running FastSurfer..."
+          echo "HAS_FASTSURFER=true" >> $GITHUB_OUTPUT
+          exit 1
+        fi
```
I dont think this is "best" ... subject ids (subject1) should not be hard coded. Also, we do not wont the runs just sometimes, we want them always.

dkuegler commented on .github/workflows/quicktest_runner.yaml:
```diff
@@ -0,0 +1,154 @@
+        echo "Running FastSurfer..."
+        export FREESURFER_HOME=/freesurfer
+        source $FREESURFER_HOME/SetUpFreeSurfer.sh
+        export FASTSURFER_HOME=$(pwd)
```
```suggestion
        export FASTSURFER_HOME=$(pwd)
        export THIS_RUN_OUTDIR=$RUNNER_FS_OUTPUT/${GITHUB_SHA:0:7}
        mkdir -p $THIS_RUN_OUTDIR
```

dkuegler commented on .github/workflows/quicktest_runner.yaml:
```diff
@@ -0,0 +1,154 @@
+        restore-keys: |
+          ${{ runner.os }}-pip-
+    - name: Run pytest
+      run: python -m pytest test/quick_test
```
This probably needs to get passed $THIS_RUN_OUTDIR

dkuegler commented on test/quick_test/common.py:
```diff
@@ -0,0 +1,46 @@
+    """
+
+    subjects_dir = os.environ["SUBJECTS_DIR"]
+    reference_id = os.environ["REFERENCE_ID"]
```
Dont use REFERENCE_ID

dkuegler commented:
Please also apply changes from the reviews and clean up the git history.

taha-abdullah commented:
@dkuegler please review

taha-abdullah committed:
Previous PR review changes

- documentation fixes in quicktest.yaml

- docstring fixes in test_errors_in_logfiles.py

- removed loop over assert in test_file_existence.py

taha-abdullah committed:
Moving Fastsurfer and Pytest environment setup to local_runner image, thus simplifying, and streamlining the workflow and improving execution time significantly.

Adding Labelled output directories for each run

Removing unnecessary import statements from pytest files

Cleaning up code:

- migrating from os.path.join to pathlib Path

- changing str to Path

- cleaning up comments and docstrings

- ran ruff on code to fix code style

taha-abdullah commented:
@dkuegler please review again

dkuegler commented on .github/workflows/quicktest_runner.yaml:
```diff
@@ -0,0 +1,87 @@
+    - name: Run pytest
+      run: |
+        source /venv-pytest/bin/activate
+        python -m pytest test/quick_test
```
Should this not get the THIS_RUN_DIR variable passed here?

taha-abdullah replied:
No. because pytest use the TEST_DIR env var and I am setting TEST_DIR to THIS_RUN_DIR in the previous job

dkuegler commented on test/quick_test/test_errors_in_logfiles.py:
```diff
@@ -0,0 +1,116 @@
+def test_errors(subjects_dir: Path, test_dir: Path, test_subject: Path):
+    """
+    Test if there are any errors in the log files.
+
```
Add documentation on the fact that subjects_dir and test_dir are filled by pytest fixtures defined in `conftest`

taha-abdullah replied:
done

dkuegler commented on test/quick_test/test_file_existence.py:
```diff
@@ -1,73 +1,71 @@
 
-    parser = argparse.ArgumentParser(description="Test for file existence based on a YAML file.")
-    parser.add_argument('folder_path', type=Path, help="The path to the folder to check.")
+    print(test_subject)
```
Clean up the print statement

taha-abdullah replied:
done

dkuegler commented on test/quick_test/test_file_existence.py:
```diff
@@ -1,73 +1,71 @@
-    This block checks if there are any command line arguments, assigns the first argument
-    to the error_file_path class variable, and runs the unittest main function.
+    Test the existence of files in the folder.
+
```
Probably also add the documentation for fixtures here.

taha-abdullah replied:
done

dkuegler commented on test/quick_test/test_images.py:
```diff
@@ -0,0 +1,222 @@
+def test_image_headers(subjects_dir: Path, test_dir: Path, reference_dir: Path, test_subject: Path):
+    """
+    Test the image headers by comparing the headers of the test and reference images.
+
```
Probably also add the documentation for the fixtures here.

taha-abdullah replied:
done

dkuegler commented on test/quick_test/test_images.py:
```diff
@@ -0,0 +1,222 @@
+
+    # Load images
+    test_subject = subjects_dir / test_dir / test_subject
+    test_image = load_image(test_subject, "brain.mgz")
```
Probably okay for now, but probably we want to test more than just one file, so how do we add more files here?

taha-abdullah replied:
working on this. I want to get the runner up and running on the server first before making these changes, if that makes sense

dkuegler commented on test/quick_test/test_images.py:
```diff
@@ -0,0 +1,222 @@
+    """
+
+    test_file = subjects_dir / test_dir / test_subject
+    test_image = load_image(test_file, "aseg.mgz")
```
Probably okay for now, but probably we want to test more than just one file, so how do we add more files here?

dkuegler commented on test/quick_test/test_images.py:
```diff
@@ -0,0 +1,222 @@
+    """
+
+    test_file = subjects_dir / test_dir / test_subject
+    test_image = load_image(test_file, "brain.mgz")
```
Probably okay for now, but probably we want to test more than just one file, so how do we add more files here?

dkuegler commented on test/quick_test/test_stats.py:
```diff
@@ -0,0 +1,267 @@
+    files = os.listdir(test_subject / "stats")
+
+    if "aseg.stats" in files:
+        return test_subject / "stats" / "aseg.stats"
```
in the end we want this test to check both of these files

taha-abdullah replied:
I would like to get the runner working on the server before making this change

dkuegler commented on test/quick_test/test_stats.py:
```diff
@@ -0,0 +1,267 @@
+        List of structs.
+    """
+
+    if "aseg" in str(test_file):
```
probably more `test_file.name == "aseg.stats"`

taha-abdullah replied:
done

taha-abdullah committed:
PR fixes:

- ruff fixes

- docstrings: added description for fixture filled arguments

dkuegler commented on test/quick_test/test_stats.py:
```diff
@@ -0,0 +1,267 @@
+                columns = line.strip("# ColHeaders").strip().split(" ")
+
+    # Read the reference table into a pandas dataframe
+    table = pd.read_table(file_path, skiprows=table_start, sep="\s+", header=None)
```
you can probably use `comment="#"` or something similar instead of this tablestart thing, but it is also fine as is.

dkuegler reviewed (changes requested):
Some minor fixes required before we merge. Generally, we need to make sure when we merge the commit, it passes the test itself.
